File: ios/Runner/Tajweed/CanonicalLexiconStore.swift
import Foundation

/// Versioned canonical spoken-Quran lexicon (ADR-010).
enum CanonicalLexiconStore {
  struct CanonicalWord: Equatable {
    let id: String
    let canonical: String
    let surfaceUthmani: String
    let surfaceIndopak: String
    let surfaceImlaei: String

    /// UI chip text — Uthmani surface when present; never used for alignment.
    var uiText: String {
      if !surfaceUthmani.isEmpty { return surfaceUthmani }
      if !surfaceImlaei.isEmpty { return surfaceImlaei }
      return canonical
    }
  }

  struct AyahLexicon: Equatable {
    let ref: String
    let words: [CanonicalWord]
  }

  struct Manifest: Equatable {
    let lexiconVersion: String
    let linguisticSpecVersion: String
    let stage2FoldRevision: String
    let ayahCount: Int
  }

  static var shadowEnabled: Bool {
    #if DEBUG
    return true
    #else
    return false
    #endif
  }

  private static var manifest: Manifest?
  private static var ayahIndex: [String: AyahLexicon]?
  private static let loadLock = NSLock()

  static func manifestInfo() -> Manifest? {
    loadIfNeeded()
    return manifest
  }

  static func ayah(surah: Int, ayah: Int) -> AyahLexicon? {
    loadIfNeeded()
    return ayahIndex?["\(surah):\(ayah)"]
  }

  private static func bundleResource(name: String, ext: String) -> URL? {
    for sub in ["canonical_lexicon", "Tajweed/canonical_lexicon", nil] {
      if let url = Bundle.main.url(forResource: name, withExtension: ext, subdirectory: sub) {
        return url
      }
    }
    return nil
  }

  static func loadIfNeeded() {
    loadLock.lock()
    defer { loadLock.unlock() }
    if ayahIndex != nil { return }

    let t0 = CFAbsoluteTimeGetCurrent()
    guard let manifestURL = bundleResource(name: "manifest", ext: "json"),
          let ndjsonURL = bundleResource(name: "ayahs", ext: "ndjson")
    else {
      NSLog("[TajweedCanonical] lexicon bundle resources missing")
      ayahIndex = [:]
      return
    }

    do {
      let manifestData = try Data(contentsOf: manifestURL)
      let manifestJSON = try JSONSerialization.jsonObject(with: manifestData) as? [String: Any] ?? [:]
      manifest = Manifest(
        lexiconVersion: manifestJSON["lexiconVersion"] as? String ?? "",
        linguisticSpecVersion: manifestJSON["linguisticSpecVersion"] as? String ?? "",
        stage2FoldRevision: manifestJSON["stage2FoldRevision"] as? String ?? "",
        ayahCount: manifestJSON["ayahCount"] as? Int ?? 0
      )

      // Stream lines instead of loading ~12MB into one String + split (same JSON rows).
      let handle = try FileHandle(forReadingFrom: ndjsonURL)
      defer { try? handle.close() }
      var index: [String: AyahLexicon] = [:]
      index.reserveCapacity(manifest?.ayahCount ?? 6236)
      var carry = Data()
      while true {
        let chunk = handle.readData(ofLength: 64 * 1024)
        if chunk.isEmpty && carry.isEmpty { break }
        carry.append(chunk)
        while let nl = carry.firstIndex(of: UInt8(ascii: "\n")) {
          let lineData = carry.subdata(in: carry.startIndex..<nl)
          carry.removeSubrange(carry.startIndex...nl)
          if lineData.isEmpty { continue }
          guard let row = try JSONSerialization.jsonObject(with: lineData) as? [String: Any],
                let ref = row["ref"] as? String,
                let wordsRaw = row["words"] as? [[String: Any]]
          else { continue }
          let words = wordsRaw.compactMap { w -> CanonicalWord? in
            guard let id = w["id"] as? String, let canonical = w["canonical"] as? String else { return nil }
            let surfaces = w["surfaceForms"] as? [String: Any] ?? [:]
            return CanonicalWord(
              id: id,
              canonical: canonical,
              surfaceUthmani: surfaces["uthmani"] as? String ?? "",
              surfaceIndopak: surfaces["indopak"] as? String ?? "",
              surfaceImlaei: surfaces["imlaei"] as? String ?? ""
            )
          }
          index[ref] = AyahLexicon(ref: ref, words: words)
        }
        if chunk.isEmpty { break }
      }
      if !carry.isEmpty,
         let row = try JSONSerialization.jsonObject(with: carry) as? [String: Any],
         let ref = row["ref"] as? String,
         let wordsRaw = row["words"] as? [[String: Any]]
      {
        let words = wordsRaw.compactMap { w -> CanonicalWord? in
          guard let id = w["id"] as? String, let canonical = w["canonical"] as? String else { return nil }
          let surfaces = w["surfaceForms"] as? [String: Any] ?? [:]
          return CanonicalWord(
            id: id,
            canonical: canonical,
            surfaceUthmani: surfaces["uthmani"] as? String ?? "",
            surfaceIndopak: surfaces["indopak"] as? String ?? "",
            surfaceImlaei: surfaces["imlaei"] as? String ?? ""
          )
        }
        index[ref] = AyahLexicon(ref: ref, words: words)
      }
      ayahIndex = index
      NSLog(
        "[TajweedCanonical] loaded lexiconVersion=%@ ayahs=%d in %.0fms (must not run on launch/main)",
        manifest?.lexiconVersion ?? "",
        index.count,
        (CFAbsoluteTimeGetCurrent() - t0) * 1000
      )
    } catch {
      NSLog("[TajweedCanonical] lexicon load failed: \(error.localizedDescription)")
      ayahIndex = [:]
    }
  }

  /// Unit tests only — inject a minimal ayah map without reading the bundle.
  static func installForTesting(ayahs: [String: AyahLexicon], version: String = "test") {
    loadLock.lock()
    defer { loadLock.unlock() }
    manifest = Manifest(
      lexiconVersion: version,
      linguisticSpecVersion: "test",
      stage2FoldRevision: "test",
      ayahCount: ayahs.count
    )
    ayahIndex = ayahs
  }

  static func resetForTesting() {
    loadLock.lock()
    defer { loadLock.unlock() }
    manifest = nil
    ayahIndex = nil
  }
}

File: android/app/src/main/kotlin/com/app/deenly/deenly/tajweed/CanonicalLexiconStore.kt
package com.rnr.deenfocus.tajweed

import android.content.Context
import android.content.pm.ApplicationInfo
import android.util.Log
import org.json.JSONObject
import java.io.BufferedReader
import java.io.InputStreamReader

/** Versioned canonical spoken-Quran lexicon (ADR-010). */
object CanonicalLexiconStore {
    data class CanonicalWord(
        val id: String,
        val canonical: String,
        val surfaceUthmani: String = "",
        val surfaceIndopak: String = "",
        val surfaceImlaei: String = "",
    ) {
        /** UI chip text — Uthmani surface when present; never used for alignment. */
        val uiText: String
            get() = when {
                surfaceUthmani.isNotEmpty() -> surfaceUthmani
                surfaceImlaei.isNotEmpty() -> surfaceImlaei
                else -> canonical
            }
    }

    data class AyahLexicon(val ref: String, val words: List<CanonicalWord>)

    data class Manifest(
        val lexiconVersion: String,
        val linguisticSpecVersion: String,
        val stage2FoldRevision: String,
        val ayahCount: Int,
    )

    @Volatile
    private var appContext: Context? = null

    @Volatile
    private var manifest: Manifest? = null

    @Volatile
    private var ayahIndex: Map<String, AyahLexicon>? = null

    @Volatile
    private var debuggable = false

    fun shadowEnabled(): Boolean = debuggable

    fun manifestInfo(): Manifest? = manifest

    fun ayah(surah: Int, ayah: Int): AyahLexicon? = ayahIndex?.get("$surah:$ayah")

    /** Store application context only — does not parse the lexicon pack. */
    @Synchronized
    fun bind(context: Context) {
        appContext = context.applicationContext
        debuggable =
            (context.applicationInfo.flags and ApplicationInfo.FLAG_DEBUGGABLE) != 0
    }

    /**
     * Parse `ayahs.ndjson` (~12MB). Call only from a background worker
     * (e.g. score path) — never from Activity / FlutterEngine configure.
     */
    @Synchronized
    fun loadIfNeeded() {
        if (ayahIndex != null) return
        val context = appContext
        if (context == null) {
            Log.w("TajweedCanonical", "loadIfNeeded skipped — bind() not called")
            ayahIndex = emptyMap()
            return
        }
        val t0 = System.nanoTime()
        try {
            context.assets.open("tajweed/canonical_lexicon/manifest.json").use { stream ->
                val json = JSONObject(stream.bufferedReader().readText())
                manifest = Manifest(
                    lexiconVersion = json.optString("lexiconVersion"),
                    linguisticSpecVersion = json.optString("linguisticSpecVersion"),
                    stage2FoldRevision = json.optString("stage2FoldRevision"),
                    ayahCount = json.optInt("ayahCount"),
                )
            }
            val index = mutableMapOf<String, AyahLexicon>()
            context.assets.open("tajweed/canonical_lexicon/ayahs.ndjson").use { stream ->
                BufferedReader(InputStreamReader(stream, Charsets.UTF_8)).forEachLine { line ->
                    if (line.isBlank()) return@forEachLine
                    val row = JSONObject(line)
                    val ref = row.getString("ref")
                    val wordsJson = row.getJSONArray("words")
                    val words = buildList {
                        for (i in 0 until wordsJson.length()) {
                            val w = wordsJson.getJSONObject(i)
                            val surfaces = w.optJSONObject("surfaceForms")
                            add(
                                CanonicalWord(
                                    id = w.getString("id"),
                                    canonical = w.getString("canonical"),
                                    surfaceUthmani = surfaces?.optString("uthmani").orEmpty(),
                                    surfaceIndopak = surfaces?.optString("indopak").orEmpty(),
                                    surfaceImlaei = surfaces?.optString("imlaei").orEmpty(),
                                ),
                            )
                        }
                    }
                    index[ref] = AyahLexicon(ref, words)
                }
            }
            ayahIndex = index
            val ms = (System.nanoTime() - t0) / 1_000_000.0
            Log.i(
                "TajweedCanonical",
                "loaded lexiconVersion=${manifest?.lexiconVersion} ayahs=${index.size} in ${"%.0f".format(ms)}ms",
            )
        } catch (e: Exception) {
            Log.w("TajweedCanonical", "lexicon load failed: ${e.message}")
            ayahIndex = emptyMap()
        }
    }

    /** Unit tests only — inject a minimal ayah map without reading assets. */
    @Synchronized
    fun installForTesting(ayahs: Map<String, AyahLexicon>, version: String = "test") {
        debuggable = true
        manifest = Manifest(
            lexiconVersion = version,
            linguisticSpecVersion = "test",
            stage2FoldRevision = "test",
            ayahCount = ayahs.size,
        )
        ayahIndex = ayahs
    }

    @Synchronized
    fun resetForTesting() {
        manifest = null
        ayahIndex = null
        debuggable = false
        appContext = null
    }
}
